### ISSUE TITLE
S3 upload stuck on problematic endpoint

### ISSUE THREAD

**Issue by alvaroaleman** [improbable-eng/thanos#315]
Hey,

Just noticed that the `thanos-sidecar` running aside of Prometheus stopped uploading data to my bucket, as you can see in the logs below it worked fine till `ts=2018-04-24T` and then there are no further uploads. Any idea about a possible reason?

<details>
<summary><code>$ kubectl logs -f prometheus-0 thanos-sidecar</code></summary>
<pre><code>
[...]
level=debug ts=2018-04-22T19:02:09.877780365Z caller=delegate.go:88 component=cluster received=NotifyLeave node=01CBPN2B3NS6AYCZEP7RYK506S addr=10.44.0.5:10900
level=debug ts=2018-04-22T19:02:13.958334843Z caller=delegate.go:82 component=cluster received=NotifyJoin node=01CBPN2B3NS6AYCZEP7RYK506S addr=10.44.0.5:10900
level=info ts=2018-04-22T19:02:23.30157061Z caller=shipper.go:185 msg="upload new block" id=01CBQBMADVDXN23YXEP23RWF9B
level=debug ts=2018-04-22T19:27:11.25943331Z caller=delegate.go:88 component=cluster received=NotifyLeave node=01CBPBSHP9WNQ5F65CSV6B1B6R addr=10.32.0.7:10900
level=debug ts=2018-04-22T19:30:02.98632032Z caller=delegate.go:82 component=cluster received=NotifyJoin node=01CBQDBAHF78JWA64EMN5Z7CR7 addr=10.32.0.10:10900
level=info ts=2018-04-22T21:00:08.262562361Z caller=shipper.go:185 msg="upload new block" id=01CBQJG1MT6DV21CVT7W48SSRZ
level=debug ts=2018-04-22T22:07:55.670535278Z caller=delegate.go:82 component=cluster received=NotifyJoin node=01CBQPCBVFGQDRP9J9H0Z4WZN6 addr=10.44.0.13:10900
level=debug ts=2018-04-22T22:08:02.258273966Z caller=delegate.go:88 component=cluster received=NotifyLeave node=01CBPMX193HZHNX1CGN43AF6WR addr=10.44.0.13:10900
level=info ts=2018-04-22T23:00:08.268362702Z caller=shipper.go:185 msg="upload new block" id=01CBQSBRY4PJPNZCEQSWABSGJW
level=info ts=2018-04-23T01:00:08.264407548Z caller=shipper.go:185 msg="upload new block" id=01CBR07G675ZA9R4BAN3K1XQWE
level=info ts=2018-04-23T03:00:08.250069176Z caller=shipper.go:185 msg="upload new block" id=01CBR737EBPQCM43JBPHNHXYGF
level=info ts=2018-04-23T05:00:08.249681035Z caller=shipper.go:185 msg="upload new block" id=01CBRDYYN32Y5XNGD0BGCXR064
level=info ts=2018-04-23T07:00:08.249493197Z caller=shipper.go:185 msg="upload new block" id=01CBRMTNXVKAMFN2JZ7AB7GXN8
level=info ts=2018-04-23T09:00:08.263506396Z caller=shipper.go:185 msg="upload new block" id=01CBRVPD6PXCX5A1F970S4CF4S
level=info ts=2018-04-23T11:00:08.262778654Z caller=shipper.go:185 msg="upload new block" id=01CBS2J4DG6TSKYKFA3XNCSEWA
level=info ts=2018-04-23T13:00:08.259102409Z caller=shipper.go:185 msg="upload new block" id=01CBS9DVPNWBRTMBM9HVJB7EKH
level=info ts=2018-04-23T15:00:08.252702363Z caller=shipper.go:185 msg="upload new block" id=01CBSG9JYPMGP9QN2AD6ERK862
level=info ts=2018-04-23T17:00:08.268642329Z caller=shipper.go:185 msg="upload new block" id=01CBSQ5A60S3EQJVASMC3XDGAN
level=info ts=2018-04-23T19:00:08.268987267Z caller=shipper.go:185 msg="upload new block" id=01CBSY11E85X2VH626FZT9G8AC
level=info ts=2018-04-23T21:00:08.268666903Z caller=shipper.go:185 msg="upload new block" id=01CBT4WRPCD0NDYHEX4BMDS085
level=info ts=2018-04-23T23:00:08.254750674Z caller=shipper.go:185 msg="upload new block" id=01CBTBRFX1E883JSG86C9F78BR
level=info ts=2018-04-24T01:00:08.254895102Z caller=shipper.go:185 msg="upload new block" id=01CBTJM75YCVVCKDKRATMSH0AN
level=info ts=2018-04-24T03:00:08.251166852Z caller=shipper.go:185 msg="upload new block" id=01CBTSFYE6Q2E52D226ACVDFXM
level=info ts=2018-04-24T05:00:08.25338784Z caller=shipper.go:185 msg="upload new block" id=01CBV0BNNKBCSEKAYV6PJAV58E
level=info ts=2018-04-24T07:00:08.261943147Z caller=shipper.go:185 msg="upload new block" id=01CBV77CWVXAS5XG7A95WG1RP0
level=info ts=2018-04-24T09:00:08.25141643Z caller=shipper.go:185 msg="upload new block" id=01CBVE3468KJWHCKKHA79J5D57
level=info ts=2018-04-24T11:00:08.24565933Z caller=shipper.go:185 msg="upload new block" id=01CBVMYVE53HWC7E3VZB4ZMVVB
level=info ts=2018-04-24T13:00:08.26298306Z caller=shipper.go:185 msg="upload new block" id=01CBVVTJPBYHWNN28MF7PMYM7G
level=info ts=2018-04-24T15:00:08.270116517Z caller=shipper.go:185 msg="upload new block" id=01CBW2P9X0KN32B5C7V2YBG0JY
level=info ts=2018-04-24T17:00:08.265739426Z caller=shipper.go:185 msg="upload new block" id=01CBW9J15Z3DQ7M9JPRN538HA4
level=info ts=2018-04-24T19:00:08.264905926Z caller=shipper.go:185 msg="upload new block" id=01CBWGDREDJ1HT1RW3Y7TM85ZD
level=debug ts=2018-04-24T20:52:19.969897464Z caller=reloader.go:97 component=reloader msg="received watch event" op=CHMOD name=/etc/prometheus/prometheus.yml.tmpl
level=info ts=2018-04-24T20:52:23.657332811Z caller=reloader.go:182 component=reloader msg="Prometheus reload triggered" cfg_in=/etc/prometheus/prometheus.yml.tmpl cfg_out=/etc/prometheus-shared/prometheus.yml rule_dir=
level=debug ts=2018-04-24T20:52:23.657560516Z caller=reloader.go:97 component=reloader msg="received watch event" op=REMOVE name=/etc/prometheus/prometheus.yml.tmpl
level=info ts=2018-04-24T21:00:08.273922735Z caller=shipper.go:185 msg="upload new block" id=01CBWQ9HRJEP82H715Q6DW95NA
level=warn ts=2018-04-26T03:25:00.556311544Z caller=sidecar.go:184 msg="heartbeat failed" err="request config against http://127.0.0.1:9090/api/v1/status/config: Get http://127.0.0.1:9090/api/v1/status/config: context deadline exceeded"
</code></pre></details>
<details>
  <summary><code>$ kubectl logs -f prometheus-1 thanos-sidecar</code></summary>
<pre> <code>
[...]
level=info ts=2018-04-22T19:02:02.846680254Z caller=shipper.go:185 msg="upload new block" id=01CBQBMADZ4ED3SE5EMA4K1FHZ
level=debug ts=2018-04-22T19:27:10.798873207Z caller=delegate.go:88 component=cluster received=NotifyLeave node=01CBPBSHP9WNQ5F65CSV6B1B6R addr=10.32.0.7:10900
level=debug ts=2018-04-22T19:30:03.000205304Z caller=delegate.go:82 component=cluster received=NotifyJoin node=01CBQDBAHF78JWA64EMN5Z7CR7 addr=10.32.0.10:10900
level=info ts=2018-04-22T21:00:12.845165248Z caller=shipper.go:185 msg="upload new block" id=01CBQJG1NZ4G1QKX1B4SBDNQNX
level=debug ts=2018-04-22T22:07:55.665867752Z caller=delegate.go:82 component=cluster received=NotifyJoin node=01CBQPCBVFGQDRP9J9H0Z4WZN6 addr=10.44.0.13:10900
level=debug ts=2018-04-22T22:08:03.258266357Z caller=delegate.go:88 component=cluster received=NotifyLeave node=01CBPMX193HZHNX1CGN43AF6WR addr=10.44.0.13:10900
level=info ts=2018-04-22T23:00:12.82464774Z caller=shipper.go:185 msg="upload new block" id=01CBQSBRY372F406Z74EMSRNAG
level=info ts=2018-04-23T01:00:12.81331659Z caller=shipper.go:185 msg="upload new block" id=01CBR07G5ZBJYQAXVBPF63R2K2
level=info ts=2018-04-23T03:00:12.818021853Z caller=shipper.go:185 msg="upload new block" id=01CBR737E1ZTB1GGGAP3MZ51NK
level=info ts=2018-04-23T05:00:12.822147111Z caller=shipper.go:185 msg="upload new block" id=01CBRDYYNZS4M0M176TBN27YW2
level=info ts=2018-04-23T07:00:12.816709604Z caller=shipper.go:185 msg="upload new block" id=01CBRMTNY236NMDPQH2DWT8TTX
level=info ts=2018-04-23T09:00:12.819262493Z caller=shipper.go:185 msg="upload new block" id=01CBRVPD5Z5QAD123XZJR2JKAG
level=info ts=2018-04-23T11:00:12.811840627Z caller=shipper.go:185 msg="upload new block" id=01CBS2J4DZ99VTF30JZ5CC698C
level=info ts=2018-04-23T13:00:12.811960581Z caller=shipper.go:185 msg="upload new block" id=01CBS9DVP04RF4V34X33J59SCZ
level=info ts=2018-04-23T15:00:12.834800274Z caller=shipper.go:185 msg="upload new block" id=01CBSG9JXX3NR87Q0S4738HK3Q
level=info ts=2018-04-23T17:00:12.8146374Z caller=shipper.go:185 msg="upload new block" id=01CBSQ5A5W1GGJKGYCXWHWH6T3
level=info ts=2018-04-23T19:00:12.80904992Z caller=shipper.go:185 msg="upload new block" id=01CBSY11DWRK256MWG7MPMD4B7
level=info ts=2018-04-23T21:00:12.813481371Z caller=shipper.go:185 msg="upload new block" id=01CBT4WRNXHJXVZZR0VE6CEYZD
level=info ts=2018-04-23T23:00:12.811728879Z caller=shipper.go:185 msg="upload new block" id=01CBTBRFY0DD2848MG193R604H
level=info ts=2018-04-24T01:00:12.812955639Z caller=shipper.go:185 msg="upload new block" id=01CBTJM75ZXKFD5772Y7PDA8GN
level=info ts=2018-04-24T03:00:12.826440544Z caller=shipper.go:185 msg="upload new block" id=01CBTSFYE31441YEBR1BNFZ8SA
level=info ts=2018-04-24T05:00:12.818482979Z caller=shipper.go:185 msg="upload new block" id=01CBV0BNNW61NV9F3CN4A3R3B4
level=info ts=2018-04-24T07:00:12.820335187Z caller=shipper.go:185 msg="upload new block" id=01CBV77CXZ234VZAC0VFJXM9QE
level=info ts=2018-04-24T09:00:12.811308843Z caller=shipper.go:185 msg="upload new block" id=01CBVE345YH5GPBHC2V1E626G3
level=info ts=2018-04-24T11:00:12.811411524Z caller=shipper.go:185 msg="upload new block" id=01CBVMYVDZ29RMW8B84RE2634Q
level=info ts=2018-04-24T13:00:12.810545342Z caller=shipper.go:185 msg="upload new block" id=01CBVVTJP1008ZENHQYGWWS8NJ
level=info ts=2018-04-24T15:00:12.817140055Z caller=shipper.go:185 msg="upload new block" id=01CBW2P9XZ6HBDJKC58GH8HKGG
level=info ts=2018-04-24T17:00:12.825472375Z caller=shipper.go:185 msg="upload new block" id=01CBW9J15VJ4DK3JQ4VFR67S55
level=info ts=2018-04-24T19:00:12.833477937Z caller=shipper.go:185 msg="upload new block" id=01CBWGDRE33QPAQEG1XAPADT7Q
level=debug ts=2018-04-24T20:52:09.106170006Z caller=reloader.go:97 component=reloader msg="received watch event" op=CHMOD name=/etc/prometheus/prometheus.yml.tmpl
level=info ts=2018-04-24T20:52:11.748298162Z caller=reloader.go:182 component=reloader msg="Prometheus reload triggered" cfg_in=/etc/prometheus/prometheus.yml.tmpl cfg_out=/etc/prometheus-shared/prometheus.yml rule_dir=
level=debug ts=2018-04-24T20:52:11.749042025Z caller=reloader.go:97 component=reloader msg="received watch event" op=REMOVE name=/etc/prometheus/prometheus.yml.tmpl
level=info ts=2018-04-24T21:00:12.828329409Z caller=shipper.go:185 msg="upload new block" id=01CBWQ9GFGJNEYT65348GPNH66
level=info ts=2018-04-24T23:00:12.82032818Z caller=shipper.go:185 msg="upload new block" id=01CBWY57Q2PGD2R6Y269J0FDFA
level=debug ts=2018-04-26T03:25:10.3578402Z caller=delegate.go:88 component=cluster received=NotifyLeave node=01CBPN33YBY5P1MDGY11AAGAQG addr=10.36.0.2:10900
level=debug ts=2018-04-26T03:26:14.459249686Z caller=delegate.go:88 component=cluster received=NotifyLeave node=01CBPBSA93MHJETH0BT9Y3V6SC addr=10.36.0.3:10900
level=debug ts=2018-04-26T03:26:26.006506074Z caller=delegate.go:82 component=cluster received=NotifyJoin node=01CBPN33YBY5P1MDGY11AAGAQG addr=10.36.0.2:10900
level=debug ts=2018-04-26T03:26:28.79744641Z caller=delegate.go:82 component=cluster received=NotifyJoin node=01CBPBSA93MHJETH0BT9Y3V6SC addr=10.36.0.3:10900
</code></pre>
</details>

**Comment by Bplotka** [comment]
Hey @alvaroaleman 

From this:
```
level=warn ts=2018-04-26T03:25:00.556311544Z caller=sidecar.go:184 msg="heartbeat failed" err="request config against http://127.0.0.1:9090/api/v1/status/config: Get http://127.0.0.1:9090/api/v1/status/config: context deadline exceeded"
``` - I see that from sidecar perspective Prometheus server has died. Can you double check that your prometheus-0 is not crash looping (because of disk out of space etc?), sharing your Prometheus logs for that time would be useful as well. (:

**Comment by Bplotka** [comment]
Are you on k8s? Similar statefulset manifest to the one in examples?

Ok, so definitely Prom seems to be fine. Somehow, both sidecar replicas stop uploading around 04-24, one at 21:00:00, second at 23:00:00 - this is kind of strange coincidence.

The context deadline on replica 0 looks like temp problem with connection to Prometheus, not really helpful fact.

Can you exec to both containera (prom and sidecar)  and check if any new block is visible for both sidecar and prom? (dirs are shared?) Maybe some change/ k8s issue prevent blocks being shared between containers? 

Cannot see anything the logic you observed in implementation, so I it might be something with your environment. Thanos uploads every blog which is present but not seen in bucket unless it is already compacted (Compaction.Level <= 1)

I could see also, that before the issue, the prometheus config was reloaded (configmap changed?). Not sure how bad config should be to make Prometheus stop putting any block (or putting blocks that are somehow empty / malformed and thus not uploadable by sidecar, since we have Prom log saying that some block was written every 2h), but worth to look on new block's meta files. Can you see anything wrong there?

**Comment by alvaroaleman** [comment]
Yeah I am on k8s, statefulset is mostly identical to the one in the examples:

<details><summary><code>prometheus-statefulset.yaml</code></summary><p>
<pre><code>
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: prometheus
  labels:
    app: prometheus
  namespace: monitoring
spec:
  replicas: 2
  selector:
    matchLabels:
      app: prometheus
      thanos-peer: "true"
  serviceName: prometheus
  template:
    metadata:
      labels:
        app: prometheus
        thanos-peer: "true"
    spec:
      serviceAccountName: prometheus
      nodeSelector:
        prometheus-host: "true"
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app
                operator: In
                values:
                - prometheus
            topologyKey: kubernetes.io/hostname
      initContainers:
      - name: fix-fs-perms
        image: busybox
        command:
          - sh
          - -c
          - 'chown -R 65534:65534 /prometheus'
        volumeMounts:
          - name: data-volume
            mountPath: /prometheus
      containers:
      - name: prometheus
        image: quay.io/prometheus/prometheus:v2.2.1
        args:
        - "--config.file=/etc/prometheus-shared/prometheus.yml"
        - "--web.external-url=https://my.url"
        - "--web.enable-lifecycle"
        - "--storage.tsdb.retention=24h"
        - "--storage.tsdb.min-block-duration=2h"
        - "--storage.tsdb.max-block-duration=2h"
        - "--storage.tsdb.path=/prometheus/data"
        ports:
        - name: web
          containerPort: 9090
        volumeMounts:
        - name: config-volume
          mountPath: /etc/prometheus
        - name: config-shared
          mountPath: /etc/prometheus-shared
        - name: data-volume
          mountPath: /prometheus
      - name: thanos-sidecar
        image: improbable/thanos:master
        env:
        - name: S3_BUCKET
          value: thanos
        - name: S3_ENDPOINT
          value: rook-ceph-rgw-s3.rook-system.svc.cluster.local
        - name: S3_ACCESS_KEY
          valueFrom:
            secretKeyRef:
              name: thanos-s3
              key: access_key
        - name: S3_SECRET_KEY
          valueFrom:
            secretKeyRef:
              name: thanos-s3
              key: secret_key
        - name: S3_INSECURE
          value: "true"
        args:
        - "sidecar"
        - "--log.level=debug"
        - "--tsdb.path=/prometheus/data"
        - "--prometheus.url=http://127.0.0.1:9090"
        - "--cluster.peers=thanos-peers.monitoring.svc.cluster.local:10900"
        - "--reloader.config-file=/etc/prometheus/prometheus.yml.tmpl"
        - "--reloader.config-envsubst-file=/etc/prometheus-shared/prometheus.yml"
        ports:
        - name: http
          containerPort: 10902
        - name: grpc
          containerPort: 10901
        - name: cluster
          containerPort: 10900
        volumeMounts:
        - name: data-volume
          mountPath: /prometheus
        - name: config-shared
          mountPath: /etc/prometheus-shared
        - name: config-volume
          mountPath: /etc/prometheus
      volumes:
      - name: config-shared
        emptyDir: {}
      - name: config-volume
        configMap:
          name: prometheus
      - name: data-volume
        hostPath:
          path: /srv/prometheus
</code></pre></details>

<br>

The files are definitely visible for the sidecar and there are new ones:

```
$ kubectl exec -it -c thanos-sidecar prometheus-0 -- sh -c 'ls -l /prometheus/data'
total 68
drwxr-xr-x    3 nobody   nogroup       4096 Apr 28 07:00 01CC5GTB0QQ9793YVKN7DV0DFX
drwxr-xr-x    3 nobody   nogroup       4096 Apr 28 09:00 01CC5QP28NXH91DPM8QHYQZW3V
drwxr-xr-x    3 nobody   nogroup       4096 Apr 28 11:00 01CC5YHSH9A0W2NXJFQAQZ1FM6
drwxr-xr-x    3 nobody   nogroup       4096 Apr 28 13:00 01CC65DGRVDR1PAR5WHVTZFC32
drwxr-xr-x    3 nobody   nogroup       4096 Apr 28 15:00 01CC6C980CAAXWEYDZAJWSBVRK
drwxr-xr-x    3 nobody   nogroup       4096 Apr 28 17:00 01CC6K4Z8TYKTXYS6Q85W687HC
drwxr-xr-x    3 nobody   nogroup       4096 Apr 28 19:00 01CC6T0MZCRKDYGDR4VJ48Z59C
drwxr-xr-x    3 nobody   nogroup       4096 Apr 28 21:00 01CC70WC7GXES9HEQHQPSDQCAS
drwxr-xr-x    3 nobody   nogroup       4096 Apr 28 23:00 01CC77R3G3FY8X8VM3NXPNZGAS
drwxr-xr-x    3 nobody   nogroup       4096 Apr 29 01:00 01CC7EKTQKY0QW1W9NBMAF8Z5P
drwxr-xr-x    3 nobody   nogroup       4096 Apr 29 03:00 01CC7NFHZJD7KD0SZB672HH7X4
drwxr-xr-x    3 nobody   nogroup       4096 Apr 29 05:00 01CC7WB97RSM3D2N4XEFM7NB4Q
drwxr-xr-x    3 nobody   nogroup       4096 Apr 29 07:00 01CC8370FK9FBMJH54BFBN8AEM
-rw-------    1 nobody   nogroup          2 Apr 29 08:30 lock
drwxr-xr-x    3 nobody   nogroup       4096 Apr 22 12:13 thanos
-rw-r--r--    1 root     root           452 Apr 24 20:59 thanos.shipper.json
drwxr-xr-x    2 nobody   nogroup       4096 Apr 29 08:30 w
```

A sample `meta.json` looks like this:

```
$ kubectl exec -it -c thanos-sidecar prometheus-0 -- sh -c 'cat /prometheus/data/01CC5GTB0QQ9793YVKN7DV0DFX/meta.json'
{
	"ulid": "01CC5GTB0QQ9793YVKN7DV0DFX",
	"minTime": 1524888000000,
	"maxTime": 1524895200000,
	"stats": {
		"numSamples": 18302408,
		"numSeries": 38131,
		"numChunks": 152521
	},
	"compaction": {
		"level": 1,
		"sources": [
			"01CC5GTB0QQ9793YVKN7DV0DFX"
		]
	},
	"version": 1
}
```

The configmap change that triggered the reload was this one:

```
@@ -416,6 +416,11 @@ data:
       - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scrape]
         action: keep
         regex: true
+      # Only keep pods that have a rfc1918 address, because the rest has
+      # hostNetwork: true and can not be scraped because of iptables rules
+      - source_labels: [__address__]
+        action: keep
+        regex: 10\..*
       - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_path]
         action: replace
         target_label: __metrics_path_
```

As a sidenote of this, what do you think about adding a metric like `thanos_objstore_bucket_last_successfull_upload_timestamp` to be able to alert on such occurences?

**Comment by Bplotka** [comment]
Metric makes sense!

Thanks for double checking all... I am currently on short holidays but will try to rethink in the evenings what could go wrong here. Nothing obvious. ):

**Comment by Bplotka** [comment]
Can you show `thanos.shipper.json`?

**Comment by alvaroaleman** [comment]
```
{
	"version": 1,
	"uploaded": [
		"01CBSY11E85X2VH626FZT9G8AC",
		"01CBT4WRPCD0NDYHEX4BMDS085",
		"01CBTBRFX1E883JSG86C9F78BR",
		"01CBTJM75YCVVCKDKRATMSH0AN",
		"01CBTSFYE6Q2E52D226ACVDFXM",
		"01CBV0BNNKBCSEKAYV6PJAV58E",
		"01CBV77CWVXAS5XG7A95WG1RP0",
		"01CBVE3468KJWHCKKHA79J5D57",
		"01CBVMYVE53HWC7E3VZB4ZMVVB",
		"01CBVVTJPBYHWNN28MF7PMYM7G",
		"01CBW2P9X0KN32B5C7V2YBG0JY",
		"01CBW9J15Z3DQ7M9JPRN538HA4",
		"01CBWGDREDJ1HT1RW3Y7TM85ZD"
	]
}
```

**Comment by Bplotka** [comment]
Thanks!

From what you gave me, other potential issues that could happen:

1. Super unlikely, but what if you already have these blocks (e.g `01CC5GTB0QQ9793YVKN7DV0DFX`) in the bucket somehow, another sidecar somewhere for same Prometheus?) 

2.More likely: I found it interesting that we can see the log line (prom 0):
`level=info ts=2018-04-24T21:00:08.273922735Z caller=shipper.go:185 msg="upload new block" id=01CBWQ9HRJEP82H715Q6DW95NA`

...as the last "uploaded" block, but can you check if it is actually uploaded? There is no such ulid in "Uploaded` thanos.shipper meta which indicates shipper being stuck somewhere.

Unfortunately, there is no metric that would increase on every shipper `Sync` (todo!), nor successful upload (there is succesful/failed objstore file upload metric though), but I think (based on your logs) we can assume so. Based on your input there are two places where it can be stuck. 
Somewhere here:
```
level.Info(s.logger).Log("msg", "upload new block", "id", meta.ULID)

	// We hard-link the files into a temporary upload directory so we are not affected
	// by other operations happening against the TSDB directory.
	updir := filepath.Join(s.dir, "thanos", "upload", meta.ULID.String())

	if err := os.RemoveAll(updir); err != nil {
		return errors.Wrap(err, "clean upload directory")
	}
	if err := os.MkdirAll(updir, 0777); err != nil {
		return errors.Wrap(err, "create upload dir")
	}
	defer os.RemoveAll(updir)

	if err := hardlinkBlock(dir, updir); err != nil {
		return errors.Wrap(err, "hard link block")
	}
	// Attach current labels and write a new meta file with Thanos extensions.
	if lset := s.labels(); lset != nil {
		meta.Thanos.Labels = lset.Map()
	}
	if err := block.WriteMetaFile(updir, meta); err != nil {
		return errors.Wrap(err, "write meta file")
	}
	return block.Upload(ctx, s.bucket, updir)
```
Or here: `ok, err := s.bucket.Exists(ctx, path.Join(meta.ULID.String(), block.MetaFilename))` for the next block.
Unless your filesystem is in failure mode, maybe S3 client does not add the reasonable timeouts. At Improbable, we use GCS only, so S3 is not that battle-tested.

Can we actually profile sidecar in this wrong state? We could list all the go routines and check what it is currently doing within sync loop, can we do it?

Sorry for delayed response, I am still technically on holidays (:

**Comment by Bplotka** [comment]
I double checked minio for retries & timeouts and there are some in place, so no obvious place for this operation being stucl

**Comment by alvaroaleman** [comment]
>but can you check if it is actually uploaded

That block was not uploaded, there is a folder named like it but its empty

So after going through the code of Thanos a little I figured the only explanation why it does not work must be in the uploading part, so I wrote a little patch just for testing:

```
 
                s := shipper.New(logger, nil, dataDir, bkt, externalLabels.Get)
 
-               ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+               ctx, cancel := context.WithCancel(context.Background())
+               ctxTO, _ := context.WithTimeout(context.Background(), 30*time.Second)
 
                g.Add(func() error {
                        defer closeFn()
 
                        return runutil.Repeat(30*time.Second, ctx.Done(), func() error {
-                               s.Sync(ctx)
+                               s.Sync(ctxTO)
 
                                minTime, _, err := s.Timestamps()
                                if err != nil {
```

The result:

```
level=info ts=2018-05-01T21:52:29.799955272Z caller=shipper.go:185 msg="upload new block" id=01CCE3FAKZVDKDWK3KN4X4EHVQ
level=error ts=2018-05-01T21:52:29.840888746Z caller=shipper.go:152 msg="shipping failed" block=01CCE3FAKZVDKDWK3KN4X4EHVQ err="upload chunks: upload file /prometheus/data/thanos/upload/01CCE3FAKZVDKDWK3KN4X4EHVQ/chunks/000001 as 01CCE3FAKZVDKDWK3KN4X4EHVQ/chunks/000001: upload s3 object: Post http://rook-ceph-rgw-s3.rook-system.svc.cluster.local/thanos/01CCE3FAKZVDKDWK3KN4X4EHVQ/chunks/000001?uploads=: context deadline exceeded"
level=info ts=2018-05-01T21:52:29.84649202Z caller=shipper.go:185 msg="upload new block" id=01CCEAB1W89JZQC4C9KKYSDAHW
level=error ts=2018-05-01T21:52:29.873434382Z caller=shipper.go:152 msg="shipping failed" block=01CCEAB1W89JZQC4C9KKYSDAHW err="upload chunks: upload file /prometheus/data/thanos/upload/01CCEAB1W89JZQC4C9KKYSDAHW/chunks/000001 as 01CCEAB1W89JZQC4C9KKYSDAHW/chunks/000001: upload s3 object: Post http://rook-ceph-rgw-s3.rook-system.svc.cluster.local/thanos/01CCEAB1W89JZQC4C9KKYSDAHW/chunks/000001?uploads=: context deadline exceeded"
level=info ts=2018-05-01T21:52:29.87726639Z caller=shipper.go:185 msg="upload new block" id=01CCEH6S4PGY41ER0APKVNGH7S
level=error ts=2018-05-01T21:52:29.9201107Z caller=shipper.go:152 msg="shipping failed" block=01CCEH6S4PGY41ER0APKVNGH7S err="upload chunks: upload file /prometheus/data/thanos/upload/01CCEH6S4PGY41ER0APKVNGH7S/chunks/000001 as 01CCEH6S4PGY41ER0APKVNGH7S/chunks/000001: upload s3 object: Post http://rook-ceph-rgw-s3.rook-system.svc.cluster.local/thanos/01CCEH6S4PGY41ER0APKVNGH7S/chunks/000001?uploads=: context deadline exceeded"
level=info ts=2018-05-01T21:52:29.922937298Z caller=shipper.go:185 msg="upload new block" id=01CCER2GC8Q8BA2T1BR5GT8BGG
level=error ts=2018-05-01T21:52:29.948412592Z caller=shipper.go:152 msg="shipping failed" block=01CCER2GC8Q8BA2T1BR5GT8BGG err="upload chunks: upload file /prometheus/data/thanos/upload/01CCER2GC8Q8BA2T1BR5GT8BGG/chunks/000001 as 01CCER2GC8Q8BA2T1BR5GT8BGG/chunks/000001: upload s3 object: Post http://rook-ceph-rgw-s3.rook-system.svc.cluster.local/thanos/01CCER2GC8Q8BA2T1BR5GT8BGG/chunks/000001?uploads=: context deadline exceeded"
level=debug ts=2018-05-01T21:52:55.90144271Z caller=delegate.go:88 component=cluster received=NotifyLeave node=01CCETZ1F4J02BHYGWXDZ3RFWF addr=10.44.0.5:10900
```

So the root cause is that uploading doesn't seem to finish and minio only sets timeouts on the [http clients transport](https://github.com/minio/minio-go/blob/master/transport.go) which is not of relevance if the uploading itself does not finish.

IMHO settings a timeout in the context like I did above to detect such cases is probably not a bad idea. I have however no idea if 30s is reasonable or not. What do you think?

**Comment by Bplotka** [comment]
> That block was not uploaded, there is a folder named like it but its empty

For `01CBWQ9HRJEP82H715Q6DW95NA`? Yea it proves that we are stuck on `s3 Upload`. So it is just empty directory, nothing in it?

Hmm are you sure your patch proves anything? I don't think so, you just set super short timeout so every job will be erroring. Simply, 30s is not enough 100% time. 

To decide what timeout should be or in what form (what to do after timeout) we need to have at least basic knowledge what problems we have. Why are we stuck? On what operation we are stuck? The profile (dumping page `http://<sidecar + prometheus IP>:<sidecar port>/debug/pprof/goroutine`) in the moment where the sidecar is stuck on infinite uploading could be useful. 

Cannot see any outgoing minio issues for similar issue. We definitely need to track down what is the root cause of it. The weird thing is that both sidecars seems to were stuck in similar time (+- 2h) so some bug on hanging S3 connection in minio is plausible here. 

Next question? I assume you already restarted sidecar. Is it reproducible or was just one time error?

**Comment by alvaroaleman** [comment]
> For 01CBWQ9HRJEP82H715Q6DW95NA? Yea it proves that we are stuck on s3 Upload. So it is just empty directory, nothing in it?

Yep, exactly.

> Hmm are you sure your patch proves anything? I don't think so, you just set super short timeout so every job will be erroring. Simply, 30s is not enough 100% time.

I would expect it to be enough under normal circumstances, the biggest subfolder in the data directory is `41.3M` in size.

The issue persists btw, I did restart the pods several times now and still no uploads or errors. The root cause is that the target s3 gateway I am running via rook.io in my cluster seems to be broken, but I think that scenario should ideally be handled in a more visible way.

I've attached the profile:
[goroutine.gz](https://github.com/improbable-eng/thanos/files/1968318/goroutine.gz)
